mysqltest/docker.go

```diff
@@ -0,0 +1,192 @@
+// up the Docker container, then runs os.Exit(1). Since the signal handler kills the process, any
+// other custom signal handlers that are installed may not get a chance to run.
+//
+// Docker must be installed on localhost for this to work. No environment vars are needed.
+func start(conf *config) (ConnInfo, Stopper, error) {
+	port, stopper, err := startContainer(conf)
+	if err != nil {
+		return ConnInfo{}, noOpStopper, err
+	}
+
+	return ConnInfo{
+		Username: "root",
+		Password: password,
+		Hostname: "localhost",
+		Port:     port,
+	}, stopper, nil
+}
+
+// Runs a MySQL docker container and returns the TCP port number for its MySQL port, along with a
+// cleanup function that stops the container.
+func startContainer(conf *config) (int, Stopper, error) {
+	// The following is based on the example from https://github.com/ory/dockertest#using-dockertest.
+	pool, err := dockertest.NewPool("")
+	if err != nil {
+		return 0, noOpStopper, fmt.Errorf("dockertest.NewPool(): %w", err)
+	}
+
+	// pulls an image, creates a container based on it and runs it
+	resource, err := pool.RunWithOptions(&dockertest.RunOptions{
+		Repository: "mysql",
+		Tag:        conf.mySQLVersion,
+		Env:        []string{"MYSQL_ROOT_PASSWORD=" + password},
+		Entrypoint: []string{"/usr/bin/timeout", strconv.Itoa(conf.killAfterSec), "docker-entrypoint.sh"},
+		Cmd:        []string{"mysqld"},
+	})
+	if err != nil {
+		var extraMsg string
+		switch {
+		case strings.Contains(err.Error(), "no such file"):
+			extraMsg = `. Please install docker: 
+		Instructions for gLinux: https://g3doc.corp.google.com/cloud/containers/g3doc/glinux-docker/install.md
+		Instructions for Debian: https://docs.docker.com/engine/install/debian/
+		Instructions for Mac: https://docs.docker.com/desktop/mac/install/`
+		case strings.Contains(err.Error(), "permission denied"):
+			extraMsg = `. To fix this, enable sudo-less docker container creation:
+					1. Run "sudo adduser $USER docker" to add your user to the docker group
+					2. Reboot the machine to make the group membership effective`
+		}
+		return 0, noOpStopper, fmt.Errorf("pool.Run() failed starting mysql container: %w%s", err, extraMsg)
+	}
+
+	var stopOnce sync.Once // only the first call to the stopper will have an effect.
+	stop := func() error {
+		var err error
+		stopOnce.Do(func() {
+			err = pool.Purge(resource)
+		})
+		if err != nil {
+			return fmt.Errorf("failed stopping MySQL docker container: %w", err)
+		}
+		return nil
+	}
+
+	// To get the TCP port number for the mysql server, we have to wait for the docker container to
+	// actually start, then get the mapped port number.
+	var outPort int
+
+	// Repeatedly try to connect to the container until it's up.
+	if err := pool.Retry(func() error {
+		p, err := fetchPort(resource)
+		if err != nil {
+			return err
+		}
+
+		if err := tryLogin(p); err != nil {
+			msg := fmt.Sprintf("MySQL isn't ready yet: %v", err)
+			log.Print(msg)
```

Comment:
I don't think we should assume the caller wants to use the stdlib logger. We could accept a logger interface, or we could have `OnX` functions that the user can call:

```go
if conf.OnLoginFailed != nil {
  conf.OnLoginFailed(something)
}
```